### PR TITLE
Environment Max Scale

### PR DESCRIPTION
**Pull request by zpatrick**
**What does this pull request do?**
* Adds/Changes environment model to include `MinScale`, `CurrentScale, `MaxScale`, and `InstanceType`. 
* Puts reasonable defaults into `config` package. 
* Allows `MaxScale` for environment to be set on `l0 environment create`, `l0 environment scale`, and through the terraform provider
* Changes how environment scale is printed on the CLI: `<min>:<current>:<max>`
* Miscellaneous improvements, bug fixes, etc. 

**How should this be tested?**
Unit Tests
Create/Update read environment via the CLI

**Checklist**
- [ ] Unit tests
~[ ] Smoke tests (will do in future PR)~
~[ ] System tests (will do in future PR)~
~[ ] Documentation (will do in future PR)~

### CODE REVIEW

File: cli/command/deploy.go [hunk 1]
```diff
@@ -74,6 +74,10 @@ func (d *DeployCommand) create(c *cli.Context) error {
 		DeployFile: content,
 	}
 
+	if err := req.Validate(); err != nil {
```
Comment (diemonster) [inline cli/command/deploy.go]: I recall you mentioning that name resolution for `deploy` wasn't working properly, is this related to that?

File: api/provider/aws/environment_create.go [hunk 2]
```diff
@@ -162,7 +169,13 @@ func (e *EnvironmentProvider) createLC(
 	return nil
 }
 
-func (e *EnvironmentProvider) createASG(autoScalingGroupName, launchConfigName string, minSize int64, privateSubnets []string) error {
+func (e *EnvironmentProvider) createASG(
+	autoScalingGroupName string,
```
Comment (diemonster) [inline api/provider/aws/environment_create.go]: [mega nit] `autoScalingGroupName, launchConfigName string,.....`

File: cli/command/environment.go [hunk 3]
```diff
@@ -76,19 +77,29 @@ func (e *EnvironmentCommand) Command() cli.Command {
...
 				Name:      "link",
-				Usage:     "links two environments together",
+				Usage:     "link two environments together",
 				Action:    e.link,
 				ArgsUsage: "SOURCE_ENVIRONMENT_NAME DESTINATION_ENVIRONMENT_NAME",
 				Flags: []cli.Flag{
 					cli.BoolTFlag{
-						Name:  FLAG_BI_DIRECTIONAL,
+						Name:  "bi-directional",
 						Usage: "specifies whether the link should be direcional",
```
Comment (diemonster) [inline cli/command/environment.go]: [typo] "direcional". Should be `bi-directional`

File: cli/printer/text_test.go [hunk 4]
```diff
@@ -43,26 +43,30 @@ func ExampleTextPrintEnvironments() {
...
 		},
 	}
 
 	printer.PrintEnvironments(environments...)
 	// Output:
-	// ENVIRONMENT ID  ENVIRONMENT NAME  OS       CLUSTER COUNT  INSTANCE SIZE  LINKS
-	// id1             name1             linux    1              m3.medium      id2
-	// id2             name2             windows  2              m3.xlarge      id1
-	//                                                                          api
+	// ENVIRONMENT ID  ENVIRONMENT NAME  OS       SCALE   INSTANCE TYPE  LINKS
```
Comment (diemonster) [inline cli/printer/text_test.go]: Seems like `SCALE` should have some sort of key for the colon separated values.

File: api/controllers/environment_test.go [hunk 5]
```diff
@@ -22,8 +22,9 @@ func TestCreateEnvironment(t *testing.T) {
 
 	req := models.CreateEnvironmentRequest{
 		EnvironmentName: "env",
-		InstanceSize:    "m3.medium",
-		MinClusterCount: 1,
+		InstanceType:    "m3.medium",
```
Comment (diemonster) [inline api/controllers/environment_test.go]: [nit] we're using a lot of different instance types/sizes throughout this changeset. I see `t2`, `m3` --- the size differences make sense in tests and such, but maybe we should consider just sticking with `t2` for type.

Comment (zpatrick) [reply api/controllers/environment_test.go]: the only non-test value is `DefaultEnvironmentInstanceType = "t2.small"` in config/defaults.go

File: api/provider/aws/environment_create.go (hunk 2 above)
Comment (zpatrick) [reply api/provider/aws/environment_create.go]: when we need to line wrap we put each param name and type on it's own line

File: cli/command/deploy.go (hunk 1 above)
Comment (zpatrick) [reply cli/command/deploy.go]: no

File: cli/printer/text_test.go (hunk 4 above)
Comment (zpatrick) [reply cli/printer/text_test.go]: Not sure what you mean?

File: api/provider/aws/environment_create.go (hunk 2 above)
Comment (diemonster) [reply api/provider/aws/environment_create.go]: TIL!

File: cli/printer/text_test.go (hunk 4 above)
Comment (diemonster) [reply cli/printer/text_test.go]: How does the user know what each value stands for? (eg. `1:2:50`)

File: cli/command/deploy.go (hunk 1 above)
Comment (diemonster) [reply cli/command/deploy.go]: is that change incoming for this PR then?

File: api/controllers/environment_test.go (hunk 5 above)
Comment (diemonster) [reply api/controllers/environment_test.go]: right, but why even keep `m3` around if we don't use it for a real default anywhere? Still a [nit] regardless.

File: cli/printer/text_test.go (hunk 4 above)
Comment (zpatrick) [reply cli/printer/text_test.go]: how do we do that for service deployments?

File: cli/command/deploy.go (hunk 1 above)
Comment (zpatrick) [reply cli/command/deploy.go]: I wasn't planning on it, it's marked as a todo in that refactor issue

File: cli/printer/text_test.go (hunk 4 above)
Comment (diemonster) [reply cli/printer/text_test.go]: I'm assuming you mean `loadbalancer get`, which is somewhat similar in that two ports and a protocol are presented. We don't give a ton of help there either, except for the `USAGE` notes for `addport` and `dropport`: e.g. `l0 loadbalancer addport [command options] NAME HOST_PORT:CONTAINER_PORT/PROTOCOL`

Anyway, I think a bit in `Usage`, accessible via `--help` would be good enough for the moment.

Comment (zpatrick) [reply cli/printer/text_test.go]: I actually meant `service get` where it shows something like:
```
SCALE
0/1 (1)
```

Comment (zpatrick) [reply cli/printer/text_test.go]: I believe that's currently explained in docs, in which case I don't see why this would be any different

File: api/provider/aws/test_aws/environment_create_test.go [hunk 6]
```diff
@@ -39,10 +40,11 @@ func TestEnvironmentCreate(t *testing.T) {
 
 	req := models.CreateEnvironmentRequest{
 		EnvironmentName:  "env_name",
-		InstanceSize:     "m3.small",
+		InstanceType:     "m3.small",
```
Comment (jparsons04) [inline api/provider/aws/test_aws/environment_create_test.go]: I don't believe that m3.small is a valid instance type anymore.

https://aws.amazon.com/ec2/instance-types/

File: api/provider/aws/test_aws/environment_read_test.go [hunk 7]
```diff
@@ -93,9 +95,11 @@ func TestEnvironmentRead(t *testing.T) {
 		EnvironmentID:   "env_id",
 		EnvironmentName: "env_name",
 		OperatingSystem: "linux",
-		InstanceSize:    "m3.small",
+		InstanceType:    "m3.small",
```
Comment (jparsons04) [inline api/provider/aws/test_aws/environment_read_test.go]: I don't believe that m3.small is a valid instance type anymore.

https://aws.amazon.com/ec2/instance-types/

File: plugins/terraform/layer0/data_source_layer0_environment_test.go [hunk 8]
```diff
@@ -29,8 +29,10 @@ func TestDataSourceLayer0EnvironmentRead(t *testing.T) {
 	environment := &models.Environment{
 		EnvironmentID:   "env_id",
 		EnvironmentName: "env_name",
-		ClusterCount:    2,
-		InstanceSize:    "m3.small",
+		MinScale:        1,
+		CurrentScale:    2,
+		MaxScale:        3,
+		InstanceType:    "m3.small",
```
Comment (jparsons04) [inline plugins/terraform/layer0/data_source_layer0_environment_test.go]: Ditto on m3.small

File: plugins/terraform/layer0/data_source_layer0_environment_test.go [hunk 9]
```diff
@@ -51,8 +53,10 @@ func TestDataSourceLayer0EnvironmentRead(t *testing.T) {
 
 	assert.Equal(t, "env_id", d.Id())
 	assert.Equal(t, "env_name", d.Get("name"))
-	assert.Equal(t, 2, d.Get("cluster_count"))
-	assert.Equal(t, "m3.small", d.Get("size"))
+	assert.Equal(t, 1, d.Get("min_scale"))
+	assert.Equal(t, 2, d.Get("current_scale"))
+	assert.Equal(t, 3, d.Get("max_scale"))
+	assert.Equal(t, "m3.small", d.Get("instance_type"))
```
Comment (jparsons04) [inline plugins/terraform/layer0/data_source_layer0_environment_test.go]: Ditto on m3.small
